File: iosapp1/Sokoban/Game/Direction.swift
import Foundation

/// 四向移动（与 html_app / androidapp1 的方向编码一致：U D L R）
enum Direction: CaseIterable {
    case up, down, left, right

    var dx: Int {
        switch self {
        case .left: return -1
        case .right: return 1
        default: return 0
        }
    }

    var dy: Int {
        switch self {
        case .up: return -1
        case .down: return 1
        default: return 0
        }
    }

    var code: Character {
        switch self {
        case .up: return "U"
        case .down: return "D"
        case .left: return "L"
        case .right: return "R"
        }
    }

    static func from(code: Character) -> Direction? {
        switch code.uppercased().first {
        case "U": return .up
        case "D": return .down
        case "L": return .left
        case "R": return .right
        default: return nil
        }
    }
}

File: n81app1/src/com/whj/sokoban/Direction.java
package com.whj.sokoban;

/**
 * 四向移动（与 html_app / android 的 U D L R 编码一致）。
 * <p>
 * J2ME 无 enum（CDC/Java SE 才有完整 enum 用法；MIDP 常用 int 常量）。
 */
public final class Direction {
    public static final int UP = 0;
    public static final int DOWN = 1;
    public static final int LEFT = 2;
    public static final int RIGHT = 3;

    public static final int[] DX = { 0, 0, -1, 1 };
    public static final int[] DY = { -1, 1, 0, 0 };

    private Direction() {}

    public static int fromCode(char ch) {
        switch (ch) {
            case 'U':
            case 'u':
                return UP;
            case 'D':
            case 'd':
                return DOWN;
            case 'L':
            case 'l':
                return LEFT;
            case 'R':
            case 'r':
                return RIGHT;
            default:
                return -1;
        }
    }
}
